### ISSUE TITLE
Segmentation fault from JuliaCall

### ISSUE THREAD

**Issue by rolfTurner** [Non-Contradiction/JuliaCall#44]
In answer to Changcheng's latest questions (we have been communicating by email up to now):

(1) Could you check the R used by Julia? For example, in Julia, you can type in the following:
julia> using RCall
julia> RCall.Rhome

Did that.  Got "/home/rolf/.julia/v0.6/Conda/deps/usr/lib/R"

(2) And check if the RCall.Rhome matches the position of the correct R.

Well there *is* an R there.  It appears to be a copy of the R that my "system" runs,
i.e. is /usr/local/bin/R. Same version (3.4.3) at least.
I'm not sure why this copy exists.  Probably came from my issuing the command
Pkg.add("RCall") in Julia.

(3) If RCall.Rhome is correct, do you have any startup file for R? Like any function to execute or package to load at the beginning?

Yes I have a file .Rprofile which does load packages.  But that makes "using RCall"
throw an error.  So I quit Julia, moved .Rprofile to save.Rprofile, restarted Julia
and got the results shown in (1). 

Got a *warning*, as I have previously mentioned by email, about Rcpp:

Warning: namespace ‘Rcpp’ is not available and has been replaced
by .GlobalEnv when processing object ‘.__C__Rcpp_stan_fit4model355f633685f5_file355f517a8896’

Don't understand this.  I have Rcpp installed.  (Ah!  Just realised: it's installed in own private
library, which Julia doesn't know about.)

I'll see if I can get it installed in the "public library" and see what happens then.

**Comment by Non-Contradiction** [comment]
Thanks for reporting.

It seems that the problem may be related to the fact that `JuliaCall` doesn't give the location of correct R to Julia and this is not the thing supposed to happen. And there can also be some potential problems related to R startup files. I will check more on this on Ubuntu.

**Comment by Non-Contradiction** [comment]
Okay, I think the first problem is related to updates on `RCall.jl` recently and should already be fixed in the latest `JuliaCall` master. Could you try to install the latest `JuliaCall` master in R by `devtools::install_github("Non-Contradiction/JuliaCall")` and see if the problem persists?

Next release on CRAN will be made in a few weeks.

**Comment by rolfTurner** [comment]
On 16/04/18 14:43, Changcheng Li wrote:

> Okay, I think the first problem is related to updates on |RCall.jl| 
> recently and should already be fixed in the latest |JuliaCall| master. 
> Could you try to install the latest |JuliaCall| master in R by 
> |devtools::install_github("Non-Contradiction/JuliaCall")| and see if the 
> problem persists?

OK.  I'll give that a try and report back.

<SNIP>

cheers,

Rolf
-- 
Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by Non-Contradiction** [comment]
The 
 ```
Solving environment: done

 # All requested packages already installed.
```
is displayed by `Conda.jl`, which is a Julia package manager to deal with dependencies, but in this case R is already installed, so it shouldn't be activated....

Could you try the following in Julia?
```
julia> using RCall
julia> RCall.Rhome
julia> RCall.libR
julia> RCall.validate_libR(RCall.libR)
```

**Comment by rolfTurner** [comment]
On 17/04/18 01:10, Changcheng Li wrote:
> The
> 
> |Solving environment: done # All requested packages already installed. |
> 
> is displayed by |Conda.jl|, which is a Julia package manager to deal 
> with dependencies, but in this case R is already installed, so it 
> shouldn't be activated....
> 
> Could you try the following in Julia?
> 
> |julia> using RCall julia> RCall.Rhome julia> RCall.libR julia> 
> RCall.validate_libR(RCall.libR) |

Results:

> julia> using RCall
> Warning: namespace ‘Rcpp’ is not available and has been replaced
> by .GlobalEnv when processing object ‘.__C__Rcpp_stan_fit4model355f633685f5_file355f517a8896’
> 
> julia> RCall.Rhome
> "/home/rolf/.julia/v0.6/Conda/deps/usr/lib/R"
> 
> julia> RCall.libR
> "/home/rolf/.julia/v0.6/Conda/deps/usr/lib/R/lib/libR.so"
> 
> julia> RCall.validate_libR(RCall.libR)
> true

BTW I installed Rcpp in the "system library", i.e.

     /usr/local/lib64/R/library

but still get the warning about Rcpp not being available.

cheers,

Rolf


-- 
Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by Non-Contradiction** [comment]
Okay, it seems that it still doesn't locate the correct R, so how about replacing the xxx in following line by the result of `R.home()` in R and run the code in Julia:
```
julia> ENV["R_HOME"] = xxx
julia> Pkg.build("RCall")
julia> using RCall
```

**Comment by rolfTurner** [comment]
On 17/04/18 09:15, Changcheng Li wrote:
> Okay, it seems that it still doesn't locate the correct R, so how about 
> replacing the xxx in following line by the result of |R.home()| in R and 
> run the code in Julia:
> 
> |julia> ENV["R_HOME"] = xxx julia> Pkg.build("RCall") julia> using RCall |

The results are as follows. Note that in the middle I had to quit from 
Julia and get rid of my .Rprofile (which loads my local "misc" package) 
--- I had forgotten to do this before starting Julia the first time.

Note also that after the call to Pkg.build("RCall") it still says that 
it is using the "local" version of R, in

    /home/rolf/.julia/v0.6/Conda/deps/usr/lib/R

etc.

> julia> ENV["R_HOME"] = "/usr/local/lib64/R"
> "/usr/local/lib64/R"
> 
> julia> Pkg.build("RCall")
> INFO: Building SpecialFunctions
> INFO: Building CodecZlib
> INFO: Building Conda
> INFO: Building RCall
> INFO: Installing R via Conda.  To use a different R installation, set the "R_HOME" environment variable and re-run Pkg.build("RCall").
> Warning: 'r' already in 'channels' list, moving to the top
> Solving environment: done
> 
> ## Package Plan ##
> 
>   environment location: /home/rolf/.julia/v0.6/Conda/deps/usr
> 
>   added / updated specs: 
>     - r-base
> 
> 
> The following packages will be downloaded:
> 
>     package                    |            build
>     ---------------------------|-----------------
>     conda-4.5.1                |           py27_0         1.0 MB
> 
> The following packages will be UPDATED:
> 
>     conda: 4.5.0-py27_0 --> 4.5.1-py27_0
> 
> 
> Downloading and Extracting Packages
> conda 4.5.1: ##################################################################### | 100% 
> Preparing transaction: done
> Verifying transaction: done
> Executing transaction: done
> INFO: Using R at /home/rolf/.julia/v0.6/Conda/deps/usr/lib/R and libR at /home/rolf/.julia/v0.6/Conda/deps/usr/lib/R/lib/libR.so.
> 
> julia> using RCall
> INFO: Recompiling stale cache file /home/rolf/.julia/lib/v0.6/CodecZlib.ji for module CodecZlib.
> INFO: Recompiling stale cache file /home/rolf/.julia/lib/v0.6/RCall.ji for module RCall.
> Loading required package: misc
> Error: package or namespace load failed for ‘misc’ in dyn.load(file, DLLpath = DLLpath, ...):
>  unable to load shared object '/home/rolf/Rlib/spatstat.utils/libs/spatstat.utils.so':
>   /home/rolf/Rlib/spatstat.utils/libs/spatstat.utils.so: undefined symbol: LENGTH
> Warning: namespace ‘Rcpp’ is not available and has been replaced
> by .GlobalEnv when processing object ‘.__C__Rcpp_stan_fit4model355f633685f5_file355f517a8896’
> Warning: namespace ‘Rcpp’ is not available and has been replaced
> by .GlobalEnv when processing object ‘.__C__Rcpp_stan_fit4model355f633685f5_file355f517a8896’
> 
> julia> quit()
> 3 Sorrow floats. mv .Rprofile save.Rprofile
> 4 Sorrow floats. julia
>                _
>    _       _ _(_)_     |  A fresh approach to technical computing
>   (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
>    _ _   _| |_  __ _   |  Type "?help" for help.
>   | | | | | | |/ _` |  |
>   | | |_| | | | (_| |  |  Version 0.6.2 (2017-12-13 18:08 UTC)
>  _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
> |__/                   |  x86_64-pc-linux-gnu
> 
> julia> using RCall
> Warning: namespace ‘Rcpp’ is not available and has been replaced
> by .GlobalEnv when processing object ‘.__C__Rcpp_stan_fit4model355f633685f5_file355f517a8896’
> 
> julia> 

cheers,

Rolf


-- 
Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by Non-Contradiction** [comment]
Yes, the problem is that it keeps using the local R. The R in "/usr/local/lib64/R" is not recognized.
Could you check whether the `libR.so` exists in the folder "/usr/local/lib64/R/lib" or the folder `R.home("lib")`?

**Comment by rolfTurner** [comment]
On 17/04/18 09:48, Changcheng Li wrote:
> Yes, the problem is that it keeps using the local R. The R in 
> "/usr/local/lib64/R" is not recognized.
> Could you check whether the |libR.so| exists in the folder 
> "/usr/local/lib64/R/lib" or the folder |R.home("lib")|?

The latter command gives "/usr/local/lib64/R/lib".

Anyhow, the answer is "no". It *does not* exist there:

> cd /usr/local/lib64
> find . -name libR.so -print
> # Nothing.
> find . -name "libR*" -print
> ./R/lib/libRblas.so
> ./R/lib/libRlapack.so

There *is* a libR.so in /usr/local/lib and in /usr/local/R/lib .

So perhaps I stuffed something up when installing R.  I installed from 
source using (basically; I omit some messy details):

  ./configure
  make
  sudo make install

And R always seems to work as advertised.  As usual, I don't understand.
Sorry 'bout that!

cheers,

Rolf

-- 
Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by rolfTurner** [comment]
On 17/04/18 10:20, Changcheng Li wrote:

> Then I think a quick solution to create a symbolic link of |libR.so| in 
> "/usr/local/lib64/R/lib" like
> |ln -s /usr/local/R/lib/libR.so /usr/local/lib64/R/lib|.
> 
> R itself on Ubuntu doesn't seem to necessarily need libR.so to work 
> correctly. Anyway, I will investigate more into the problem of locating 
> the libR.so on Ubuntu.

I made the suggested symbolic link and then redid the 
Pkg.build("RCall").  It still seems to talk about using the local 
version of R:

> julia> Pkg.build("RCall")
> 
> INFO: Building SpecialFunctions
> INFO: Building CodecZlib
> INFO: Building Conda
> INFO: Building RCall
> INFO: Using previously configured R at /home/rolf/.julia/v0.6/Conda/deps/usr/lib/R with libR in /home/rolf/.julia/v0.6/Conda/deps/usr/lib/R/lib/libR.so.
> 
> julia> 
> 
> julia> using RCall
> INFO: Recompiling stale cache file /home/rolf/.julia/lib/v0.6/CodecZlib.ji for module CodecZlib.
> INFO: Recompiling stale cache file /home/rolf/.julia/lib/v0.6/RCall.ji for module RCall.
> Warning: namespace ‘Rcpp’ is not available and has been replaced
> by .GlobalEnv when processing object ‘.__C__Rcpp_stan_fit4model355f633685f5_file355f517a8896’

And of course it still can't fine Rcpp.

I tried setting R_HOME in various ways before calling 
Pkg.build("RCall"); (a) by using ENV["R_HOME"] = "/usr/local/lib64/R" in 
Julia, and (b) by setting an environment variable R_HOME equal to 
"/usr/local/lib64/R".

Didn't seem to make any difference.  Exasperating!

cheers,

Rolf

-- 
Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by Non-Contradiction** [comment]
Maybe we could try to check the validity of the libR.so like this:
```
julia> using RCall
julia> RCall.validate_libR("/usr/local/R/lib/libR.so")
julia> RCall.validate_libR("/usr/local/lib/libR.so")
julia> RCall.validate_libR("/usr/local/lib64/R/lib/libR.so")
```

**Comment by rolfTurner** [comment]
On 17/04/18 12:57, Changcheng Li wrote:
> Maybe we could try to check the validity of the libR.so like this:
> 
> |julia> using RCall julia> 
> RCall.validate_libR("/usr/local/R/lib/libR.so") julia> 
> RCall.validate_libR("/usr/local/lib/libR.so") julia> 
> RCall.validate_libR("/usr/local/lib64/R/lib/libR.so") |

Sorry, I messed up.  Somehow I miss-read where the libR.so files were 
located.  They are in /usr/lib and in /usr/R/lib (the former is a 
symbolic link to the latter) and *NOT* in /usr/local/lib or
/usr/local/R/lib.

Sigh!  If there is a way to get something wrong, I will find it!

I really do apologise.

I redid the symbolic link:

sudo ln -s /usr/R/lib/libR.so /usr /usr/local/lib64/R/lib

I then re-did the Pkg.build("RCall") command in Julia and got:

> INFO: Building SpecialFunctions
> INFO: Building CodecZlib
> INFO: Building Conda
> INFO: Building RCall
> INFO: R library /usr/local/lib64/R/lib/libR.so appears to be too old. RCall.jl requires R 3.4.0 or later
> ....
So there is no *usable* libR.so on my system.'

I have little feeling for what to do about this.  A small bit of 
Googling around revealed a couple of postings indicating that one needs 
to build R using the flag --enable-R-shlib .

Can you tell me if this is what I should do?

cheers,

Rolf

-- 
Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by Non-Contradiction** [comment]
Everyone make mistakes.

Building R from source again using the flag is surely an option. It should be okay and might not be the easiest way to do. Is it necessary to build R from source?

I myself follow the steps on CRAN to install R on Ubuntu virtual machine on my laptop and it works well,  like:
```
sudo apt-get update
sudo apt-get install r-base
sudo apt-get install r-base-dev
```
but I'm not sure if this suits your need, and lots of R packages might need to reinstall...

**Comment by rolfTurner** [comment]
On 18/04/18 01:13, Changcheng Li wrote:
> I am also confused by this....
> Maybe we could open the RCall file and directly to modify it to what we 
> want,
> for example, in julia, get the location of RCall's setting file by:
> 
> |julia> Base.Pkg.dir("RCall", "deps", "deps.jl") |
> 
> then open this file by some text editor and edit it to what we want, 
> note that it defines the Rhome and libR positions used by RCall.
> And then we restart julia again, with no setting of any environment 
> variable, do |Pkg.build("RCall")| directly.

Hmm.  The software seems to be very stubborn.  I issued the command as 
directed, and got

     "/home/rolf/.julia/v0.6/RCall/deps/deps.jl"

So I edited that file, which contained the lines:

> const Rhome = "/home/rolf/.julia/v0.6/Conda/deps/usr/lib/R"
> const libR = "/home/rolf/.julia/v0.6/Conda/deps/usr/lib/R/lib/libR.so"

I changed these to

> const Rhome = "/usr/local/lib64/R"
> const libR = "/usr/local/lib64/R/lib/libR.so"

and wrote the file.

I then re-started Julia and ran Pkg.build("RCall") .

Same results as before!  I then looked at the file

     /home/rolf/.julia/v0.6/RCall/deps/deps.jl .

It was back to being the same as it had been before I edited it, i.e. it was

> const Rhome = "/home/rolf/.julia/v0.6/Conda/deps/usr/lib/R"
> const libR = "/home/rolf/.julia/v0.6/Conda/deps/usr/lib/R/lib/libR.so"

again.

The computer gods don't like me. :-)

cheers,

Rolf


Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by Non-Contradiction** [comment]
Okay.... The problem is really weird...
Let's try
```
julia> Rhome = "/usr/local/lib64/R"
julia> isdir(Rhome) ## this should be true
julia> using RCall
julia> RCall.locate_libR(Rhome) ## this should work and return "/usr/local/lib64/R/lib/libR.so"
julia> libR = joinpath(Rhome, "lib", "libR.$(Libdl.dlext)") ## should be "/usr/local/lib64/R/lib/libR.so"
julia> RCall.validate_libR(libR)
```

If all the above works, then there really shouldn't be any problem......

**Comment by rolfTurner** [comment]
On 18/04/18 10:00, Changcheng Li wrote:

> Okay.... The problem is really weird...
> Let's try

> julia> Rhome = "/usr/local/lib64/R"

Did that.

> julia> isdir(Rhome) ## this should be true

Yes, it was.

> julia> using RCall

Did that.  The response was:

> INFO: Recompiling stale cache file /home/rolf/.julia/lib/v0.6/RCall.ji for module RCall.
> Warning: namespace ‘Rcpp’ is not available and has been replaced
> by .GlobalEnv when processing object ‘.__C__Rcpp_stan_fit4model355f633685f5_file355f517a8896’

Then did:

> julia> RCall.locate_libR(Rhome) ## this should work and return "/usr/local/lib64/R/lib/libR.so"

It did work, and returned the expected string.

Then did:

> julia> libR = joinpath(Rhome, "lib", "libR.$(Libdl.dlext)") ## should be "/usr/local/lib64/R/lib/libR.so"

Worked; string returned as expected.

> julia> RCall.validate_libR(libR)

Returned "true".

> If all the above works, then there really shouldn't be any problem......

Well, all of the above worked.

So I quit from Julia, started R and did:

> library(JuliaCall)
> jh <- "/home/rolf/Desktop/Julia/julia-d386e40c17/bin/"
> system.time(j <- julia_setup(JULIA_HOME=jh,verbose=TRUE))

I got:

> Julia version 0.6.2 at location /home/rolf/Desktop/Julia/julia-d386e40c17/bin will be used.
> Julia initiation...
> Finish Julia initiation.
> Loading setup script for JuliaCall...
> Finish loading setup script for JuliaCall.
>    user  system elapsed 
>  36.444   3.196  36.370 

No seg fault!!!

I guess I hadn't tried using the JuliaCall package since re-installing R 
with the --enable-R-shlib flag set in the configure step.  Maybe this 
has fixed the problem???

I shall experiment further.

I still don't really get what's going on, but.  Too many bits and pieces 
to coordinate.  Can't keep them all straight.

cheers,

Rolf

-- 
Technical Editor ANZJS
Department of Statistics
University of Auckland
Phone: +64-9-373-7599 ext. 88276

**Comment by Non-Contradiction** [comment]
It's great to hear that `JuliaCall` starts with no segfaults!!
I will close this issue and if there is any other problem, it's totally okay to open another issue.
